src/config.rs

```diff
@@ -334,6 +334,83 @@ impl Context {
     }
 }
 
+// Separate impl block for self address handling
+impl Context {
+    /// determine whether the specified addr maps to the/a self addr
+    pub(crate) async fn is_self_addr(&self, addr: &str) -> Result<bool> {
+        let addr = addr_normalize(addr).to_lowercase();
+
+        // The addresses we get here are already normalized and lowercase
+        Ok(
+            self.get_config(Config::ConfiguredAddr).await?.as_deref() == Some(&addr)
+                || self.get_secondary_self_addrs().await?.contains(&addr),
+        )
+    }
+
+    /// Sets `primary_new` as the new primary self address and saves the old
+    /// primary address (if exists) as a secondary address.
+    pub(crate) async fn set_primary_self_addr(&self, primary_new: &str) -> Result<()> {
+        let primary_new = addr_normalize(primary_new).to_lowercase();
+
+        // add old primary address (if exists) to secondary addresses
+        let mut secondary_addrs = self.get_all_self_addrs().await?;
+
+        for a in secondary_addrs.iter_mut() {
+            *a = addr_normalize(a).to_lowercase();
+        }
+
+        // never store a primary address also as a secondary
+        secondary_addrs.retain(|a| a != &primary_new);
+        self.set_config(
+            Config::SecondaryAddrs,
+            Some(secondary_addrs.join(" ").as_str()),
+        )
+        .await?;
+
+        self.set_config(Config::ConfiguredAddr, Some(&primary_new))
+            .await?;
+
+        Ok(())
+    }
+
+    /// Returns all primary and secondary self addresses.
+    pub(crate) async fn get_all_self_addrs(&self) -> Result<Vec<String>> {
+        let mut ret = Vec::new();
+
+        ret.push(self.get_primary_self_addr().await.unwrap_or_default());
```

Comment:
This will push an empty string if the primary self addr is not set, I don't think that's what we want.

Reply:
I pushed 3f110883 which fixes this